### PR TITLE
Migration towards nnbd.

### PR DESCRIPTION
**Pull request by happy-san**
I'm working on AvlTree, I've migrated it to nnbd but I'm updating comments.
#35

### CODE REVIEW

File: lib/trees/avl_tree.dart [hunk 1]
```diff
@@ -8,31 +7,37 @@ class AvlNode<V extends Comparable> extends BinaryNodeADT<AvlNode<V>, V> {
   /// Difference between height of left and right subtree.
   ///
   /// [balanceFactor] ∈ `{-1, 0, 1}`.
-  /// Any [AvlNode] having [balanceFactor] outside this set is imbalanced.
+  /// Any [AvlNode] having [balanceFactor] that does not belong to this set is
+  /// considered imbalanced.
+  ///
+  /// In [AvlTree] it is the primary focus to actively balance out all
+  /// imbalanced [node]s following addition or deletion.
   int balanceFactor = 0;
```
Comment (happy-san) [inline lib/trees/avl_tree.dart:15]: How about using a Map here? Combines the good of both enum and current definition
'rightHeavy' -> -1
'balanced' -> 0
'leftHeavy' -> 1

File: lib/trees/avl_tree.dart [hunk 2]
```diff
@@ -348,88 +364,144 @@ class AvlTree<V extends Comparable> extends BinaryTreeADT<AvlNode<V>, V> {
...
+      //        ** **
       case 0:
-        // node is right heavy now.
+        // [node] is right heavy now.
+        //          N
+        //         / \
+        //        *  **
         node.balanceFactor = -1;
         _isShorter = false;
         break;
```
Comment (happy-san) [inline lib/trees/avl_tree.dart]: @code-shoily Do you think the added comment helps or it's just cluttering?

File: lib/trees/avl_tree.dart (hunk 1 above)
Comment (happy-san) [reply lib/trees/avl_tree.dart:15]: Actually, I like this implementation 

```dart
enum BalanceState { rightHeavy, balanced, leftHeavy }

class Node {
  BalanceState balanceState = BalanceState.balanced;

  int get balanceFactor {
    switch (balanceState) {
      case BalanceState.rightHeavy:
        return -1;
      case BalanceState.balanced:
        return 0;
      case BalanceState.leftHeavy:
        return 1;
    }
  }
}

void main() {
  var node = Node();

  print(node.balanceState);
  print(node.balanceFactor);
}
```

This prints

```
BalanceState.balanced
0
```

Comment (happy-san) [reply lib/trees/avl_tree.dart:15]: It just occured to me, if a node is leftHeavy, it's balance factor can be any natural number. Similarly rightHeavy node's balance factor can be any negative integer. So by definition, it cannot be hardcoded, like I did.

File: lib/trees/avl_tree.dart (hunk 2 above)
Comment (code-shoily) [reply lib/trees/avl_tree.dart]: No it does help.

**Review by code-shoily — APPROVED** [review]

**Pull request merged**